compendium/test_bundles/TestBundleManagedServiceDeadlock/src/ManagedServiceImpl.cpp

```diff
@@ -0,0 +1,58 @@
+#include "ManagedServiceImpl.hpp"
+
+#include <cppmicroservices/BundleContext.h>
+#include <cppmicroservices/Framework.h>
+#include <cppmicroservices/FrameworkFactory.h>
+#include <iostream>
+#include <string>
+
+namespace cppmicroservices {
+namespace service {
+namespace cm {
+namespace test {
+
+TestManagedServiceImpl2::TestManagedServiceImpl2()
+  : m_counter{ 0 }
+{}
+
+TestManagedServiceImpl2::~TestManagedServiceImpl2() = default;
+
+void TestManagedServiceImpl2::Updated(AnyMap const& props)
+{
+  {
+    std::lock_guard<std::mutex> lk(m_counterMtx);
+    m_counter++;
+  } //unlock m_counterMtx
+
+  if (!props.empty()) {
+    std::lock_guard<std::mutex> lock(m_propertiesMutex);
+
+    if (!props.AtCompoundKey("myMutex", Any()).Empty()) {
+      m_cvMtx = cppmicroservices::any_cast<std::shared_ptr<std::mutex>>(
+          props.at("myMutex"));
+    }
+   
+    if (!props.AtCompoundKey("myConditionVariable", Any()).Empty()) {
+      m_cv = cppmicroservices::any_cast<std::shared_ptr<std::condition_variable>>(
+          props.at("myConditionVariable"));
+    }
+   } // unlock m_propertiesMutex
+
+   // This is simulating a deadlock.
+   if (m_cvMtx && m_cv) {
+      std::unique_lock<std::mutex> lck(*m_cvMtx);
+      m_cv->wait(lck);
+   }
```

Comment:
I originally wrote the test that way with the Updated method getting the TestBundleManagedServiceDeadlock bundle and calling Stop. I couldn't make the problem happen with that implementation because of the changes that you made to fix the missing config updated event (PR#727). When the bundle is stopped both ConfigurationAdminImpl::RemovedService and ConfigurationAdminImpl::RemoveConfigurations get called. RemovedService gets called first and removes the service from the trackedManagedServices_ vector. When RemoveConfigurations gets called there is no service to be notified so the Removed notification isn't sent thereby avoiding the deadlock.

Reply:
I think a test case is missing then and there may be a regression caused by PR #727 related to how configuration events are sent to ManagedService[Factory] services when the bundle contains configuration data in manifest.json. 

Having a bundle that provides a configuration in it's manifest.json and doesn't implement a ManagedService nor ManagedServiceFactory will allow testing by stopping the bundle.

Reply:
I've logged issue #746 to track the regression caused by PR #727